src/util/db/syncing.ts

```diff
@@ -0,0 +1,188 @@
+  )
+
+export interface RepoUpdates {
+  checkpoints: CheckpointArray
+  changeSet: ChangeSetV2
+}
+
+export const readUpdates = (appState: AppState) => async (
+  repoId: string,
+  checkpoints: CheckpointArray
+): Promise<RepoUpdates> => {
+  const defaultCheckpoint = { version: 0, sum: 0 }
+  const currentCheckpoint = checkpoints[0] ?? defaultCheckpoint
+
+  // Query view for file documents with version > currentCheckpoint.version
+  const fileDocumentsResponse = await appState.dataStore.partitionedView(
+    repoId,
+    'versioning',
+    'version',
+    {
+      include_docs: true,
+      reduce: false,
+      // Add 1 because range is exclusive
+      start_key: currentCheckpoint.version + 1
+    }
+  )
+
+  const storeFileDocs = fileDocumentsResponse.rows.map(row => {
+    return trial(
+      () => asStoreFileDocument(row.doc),
+      err => {
+        throw new Effect(err, `Expected StoreFile document for ${row.id}`)
+      }
+    )
+  })
+
+  // Query view for checkpoint sum where version ≤ currentCheckpoint.version
+  const { sum: wipSum } = await getCheckpointAt(appState)(
+    repoId,
+    currentCheckpoint.version
+  )
+
+  // The expected difference between client and server sums
+  // caused by overwritten documents.
+  const overwriteDiff = storeFileDocs.reduce((sum, storeFile) => {
+    const historicalVersions = storeFile.versions
+      .slice(1)
+      .filter(version => version <= currentCheckpoint.version)
+
+    return sum + Math.max(0, ...historicalVersions)
+  }, 0)
+
+  // Adjusted checkpoint sum includes the overwriteDiff
+  const adjustedSum = wipSum + overwriteDiff
+
+  // Verify current checkpoint
+  // If not equal, drop highest client checkpoint and retry
+  if (adjustedSum !== currentCheckpoint.sum) {
+    return await readUpdates(appState)(repoId, checkpoints.slice(1))
+  }
+
+  // Calculate latest checkpoint
+  const latestCheckpoint = calculateLatestCheckpoint(
+    currentCheckpoint,
+    storeFileDocs
+  )
+
+  // Include latest checkpoint in new checkpoints array
+  const updatedCheckpoints = equalCheckpoints(
+    currentCheckpoint,
+    latestCheckpoint
+  )
+    ? [...checkpoints]
+    : [latestCheckpoint, ...checkpoints]
+
+  // Prepare changeSet
+  const changeSet: ChangeSetV2 = storeFileDocs.reduce(
+    (changeSet, storeFileDoc) => {
+      const path = storeFileDoc._id.split(':')[1]
+      return { ...changeSet, [path]: storeFileDoc.box }
+    },
+    {}
+  )
+
+  // Return changeset and new checkpoints
+  return {
+    checkpoints: updatedCheckpoints,
+    changeSet
+  }
+}
+
+/**
+ * Calculates what would be the latest checkpoint given set of new documents
+ * past a given checkpoint.
+ */
+const calculateLatestCheckpoint = (
+  checkpoint: Checkpoint,
+  newDocuments: StoreFileDocument[]
+): Checkpoint => {
+  // The max of all the new document versions
+  const latestVersion = newDocuments.reduce(
+    (version, document) => Math.max(version, document.versions[0]),
```

Comment:
Again, should just be first one, since they are in order.

Reply:
First document? If we assume the documents are in order, then it'd be the last one.

Reply:
Oh, I thought they were in descending order? Either way, first or last, we can just get the end one.

Reply:
It depends on the database query. So if we assume the order, the function would break if the input doesn't meet the ordering requirement. We could be a looser here and require that the documents just be sorted and do 

```ts
Math.max(newDocuments[0].version[0], newDocuments[newDocuments.length - 1].version[0])
```

Reply:
Oh, wait. `document.versions[0]` is what I wanted, and I see you are doing that. So, nevermind. I was confused.